feat: Create features and environments from extras

olivier-lacroix opened the pull request:
fixes #1062

olivier-lacroix commented:
`pixi init` works, and adds the correct `tool.pixi.environments` section to the `pyproject.toml`

`pixi install` does not, and yields `missing python interpreter from environment`. Shouldn't it be picked from the default feature? What am I missing here?

olivier-lacroix commented on src/cli/init.rs:
```diff
@@ -62,6 +64,13 @@ const PYROJECT_TEMPLATE: &str = r#"
 name = "{{ name }}"
 channels = [{%- if channels %}"{{ channels|join("\", \"") }}"{%- endif %}]
 platforms = ["{{ platforms|join("\", \"") }}"]
```
I am not sure the `join` is needed at all. The following seems to work alright
```suggestion
platforms = {{platforms}}
```

ruben-arts replied:
Makes sense! If it works lets change it.

olivier-lacroix replied:
Done in https://github.com/prefix-dev/pixi/pull/1077/commits/bbe7d9e0309fcd2051ecdf10402904a32754587e

olivier-lacroix commented on src/cli/init.rs:
```diff
@@ -62,6 +64,14 @@ const PYROJECT_TEMPLATE: &str = r#"
+{%- if loop.first %}
+
+[tool.pixi.environments]
+default = {features = [], solve-group = "default"}
```
Without the line `default = {features = [], solve-group = "default"}` `pixi install` does not work, and yields `missing python interpreter from environment`. I am not sure this is expected?

In any case, my understanding is this is the right syntax to solve all environments (including the default one) as part of the same solve-group; is that correct?

ruben-arts replied:
I would say using the `"default"` solve-group without defining it in the "default" env would make sense to me. What do you think @pavelzw @baszalmstra?

But this error would be unexpected to me, maybe @baszalmstra know's what is going on? As I would have expected an `"default" solve group doesn't exist` error.

baszalmstra replied:
Solve groups are just names you can define anywhere. The error is weird to me too, Ill check when I have the time.

ruben-arts replied:
I meant to say that we could implicitly add the `solve-group = "default"` to the default env so users don't have to define it separately. WDYT?

baszalmstra replied:
But “default” does not have a special meaning or anything? Other environments do not by default have the “default” solvegroup. What do you hope to achieve?

ruben-arts replied:
I think it makes sense to have a "default" everything. 

- Default feature
- Default environment
- Default solve group

olivier-lacroix replied:
Added Default SolveGroup in in https://github.com/prefix-dev/pixi/pull/1077/commits/96385f2eaf45090a6c78dc86684bc019bf6dc252

This will only work if the Default Environment is not redefined

olivier-lacroix replied:
Also, I could not reproduce the issue with `missing python interpreter from environment`. so seems to be all good there.

pavelzw replied:
wouldn't this imply that every environment is in the same solve group if not specifying anything? Which was one thing we originally wanted to avoid?

olivier-lacroix replied:
Currently, the only environment in the 'default' solve-group is the default environment.

olivier-lacroix commented:
No idea why the doc building fails 🙈

olivier-lacroix commented on src/project/manifest/mod.rs:
```diff
@@ -1022,7 +1033,7 @@ impl<'de> Deserialize<'de> for ProjectManifest {
                 features: Vec::new(),
                 features_source_loc: None,
-                solve_group: None,
+                solve_group: Some(solve_groups.insert("default", 0)),
```
It will only create a "default" solve-group if the default environment is not redefined. Should we make sure it is created even if the default environment is redefined?

olivier-lacroix commented on src/project/manifest/mod.rs:
```diff
@@ -1022,16 +1047,15 @@ impl<'de> Deserialize<'de> for ProjectManifest {
 
         // Add all named environments
+        // Note: if the default Environment is redefined, but does not specify a solve-group,
+        // it is not added to the 'default' solve-group.
```
Is this OK?

Or do we want to add it to the 'default' solve-group? That is, the following

```toml
[environments]
default = ["somefeature"]
```

Would be parsed as if the TOML contained

```toml
[environments]
default = { features = ["somefeature"], solve-group = "default" } 
```

olivier-lacroix commented:
@ruben-arts @baszalmstra any thought on this PR?

ruben-arts commented on docs/advanced/pyproject_toml.md:
```diff
@@ -104,6 +104,50 @@ matplotlib = "*"
 As pixi takes the conda dependencies over the pypi dependencies.
 
+## Optional dependencies
+If your python project includes groups of optional dependencies, pixi will automatically interpret them as pixi features of the same name with the associated pypi dependencies.
```
```suggestion
If your python project includes groups of optional dependencies, pixi will automatically interpret them as [pixi features](../configuration.md#the-feature-table) of the same name with the associated `pypi-dependencies`.
```

ruben-arts reviewed (approved):
I like it, it relatively simple and small but behaves like the `[project.dependencies]` which I am for. The ability to overwrite and extend using the normal pixi feature tables is also nice.

olivier-lacroix commented:
@baszalmstra there you go: https://github.com/prefix-dev/pixi/pull/1077#discussion_r1545470875

ruben-arts commented:
I'm sorry for the confusement @olivier-lacroix, @baszalmstra kinda convinced me outside of this thread. Lets put the `default = {features = [], solve-group = "default"}`  back in. It's easier to make this implicit after having it explicit then the other way around. 
As making this implicit is a breaking change. Because in theory user could already be using the "default" solve-group for non-default environments.

olivier-lacroix commented:
@ruben-arts no worries, I reverted that bit. I also:

- Refactored the parsing of `pyproject.toml` in `init.rs` to take advantage of the new custom `TomlError`
- Added an info message about the created environments if any (to align with what was done re. the install of the package itself).
- Added an info message when `pixi init` is run against a `pyproject.toml` that contains a pixi section already
- Updated the flask example

ruben-arts commented:
Thanks @olivier-lacroix !

dhirschfeld commented on docs/advanced/pyproject_toml.md:
````diff
@@ -103,6 +103,49 @@ matplotlib = "*"
+```
+In this example, three environments will be created by pixi:
+
+ - **default** with 'package1' as pypi dependency
+ - **test** with 'package1' and 'pytest' as pypi dependencies
+ - **all** with 'package1', 'package2' and 'pytest' as pypi dependencies
````
With the above, is it correct that installing the `test` environment with `pixi` would install `pytest` from `pypi` rather than `conda-forge`?

If I wanted my `pixi` `test` environment to only install dependencies from `conda-forge` can I specify that?

olivier-lacroix replied:
@dhirschfeld you are correct.

Currently, you would have to manually add them to the `tool.pixi.feature.test.dependencies` table. What you want is also on my wishlist :-) https://github.com/prefix-dev/pixi/issues/532

dhirschfeld replied:
So, if I did manually specify the test dependencies in `[tool.pixi.feature.test.dependencies]` table, then that would be used ***instead*** of mapping `pypi` dependencies from the `[project.optional-dependencies]` table?

olivier-lacroix replied:
Yes. Pixi currently solves pypi dependencies after the conda ones. And if it detects a package is already provided by one of the conda packages requested, it is skipped.
